contracts/src/mocks/MockFunctionGateway.sol

```diff
@@ -1,4 +1,184 @@
+
+/// @title MockFunctionGateway
+/// @notice A Mock version of FunctionGateway for testing.
+/// @dev This contract is only meant to be used in tests. To use it, create a json fixture with keys "input"
+///      and "output". Call loadFixture() with the path, then this contract will automatically fulfill
+///      requests with the output.
+contract MockFunctionGateway is IFunctionGateway {
+    VmSafe private constant vm = VmSafe(address(uint160(uint256(keccak256("hevm cheat code")))));
+    uint32 public nonce;
+    bytes32 public verifiedFunctionId;
+    bytes32 public verifiedInputHash;
+    bytes public verifiedOutput;
+    bool public isCallback;
+    mapping(uint32 => bytes32) public requests;
+    mapping(bytes32 => bytes) public outputs;
+
+    function loadFixture(string memory _path) external {
+        string memory json;
+
+        try vm.readFile(_path) returns (string memory data) {
+            json = data;
+        } catch {
+            revert("MockFunctionGateway: fixture not found");
+        }
+
+        bytes memory input = stdJson.readBytes(json, "$.input");
+        bytes memory output = stdJson.readBytes(json, "$.output");
+        bytes32 inputHash = sha256(input);
+        outputs[inputHash] = output;
+    }
+
+    function requestCallback(
+        bytes32 _functionId,
+        bytes memory _input,
+        bytes memory _context,
+        bytes4 _callbackSelector,
+        uint32 _callbackGasLimit
+    ) public payable returns (bytes32) {
+        bytes32 inputHash = sha256(_input);
+        bytes32 contextHash = keccak256(_context);
+        address callbackAddress = msg.sender;
+        bytes32 requestHash = _requestHash(
+            nonce,
+            _functionId,
+            inputHash,
+            contextHash,
+            callbackAddress,
+            _callbackSelector,
+            _callbackGasLimit
+        );
+
+        requests[nonce] = requestHash;
+        emit RequestCallback(
+            nonce,
+            _functionId,
+            _input,
+            _context,
+            callbackAddress,
+            _callbackSelector,
+            _callbackGasLimit,
+            msg.value
+        );
+        nonce++;
+
+        // If fixture has been pre-loaded, automatically fulfill and callback.
+        bytes memory output = outputs[inputHash];
+        if (output.length > 0) {
+            bytes32 outputHash = sha256(output);
+
+            isCallback = true;
+            (bool status,) =
+                msg.sender.call(abi.encodeWithSelector(_callbackSelector, output, _context));
+            isCallback = false;
+            if (!status) {
+                revert CallbackFailed(_callbackSelector, output, _context);
+            }
+
+            emit RequestFulfilled(nonce, _functionId, inputHash, outputHash);
+        }
+
+        return requestHash;
+    }
+
+    /// @dev This function only exists to implement interface. Use requestCall with your target
```

Comment:
Should we have it revert if it only exists to implement the interface? I'm kind of confused haha why we have this in the interface anyways actually

Reply:
In the normal case, it seems fine for `Call` to have different requesting contract and receiving contract. We just have this hear as a convince to just make it one function by assuming the requestinad and receiving are the same.